### PR TITLE
CPM-834: Refacto index product and product model

### PR DESCRIPTION
**Pull request by antoinearrive**
**Description (for Contributor and Core Developer)**

**Definition Of Done (for Core Developer only)**

- [ ] Tests
- [ ] Migration & Installer
- [ ] PM Validation (Story)
- [ ] Changelog (maintenance bug fixes)
- [ ] Tech Doc

### CODE REVIEW

File: src/Akeneo/Pim/Enrichment/Bundle/Storage/ElasticsearchAndSql/ProductAndProductModel/IndexProductAndProductModel.php [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+ */
+final class IndexProductAndProductModel
+{
+    public function __construct(
+        private readonly Client $productAndProductModelClient,
+        private readonly Connection $connection,
+    ) {
+    }
+
+    public function getDiffProductUuids(int $batchSize): iterable
```
Comment (ahocquard) [inline src/Akeneo/Pim/Enrichment/Bundle/Storage/ElasticsearchAndSql/ProductAndProductModel/IndexProductAndProductModel.php:24]:
```suggestion
    public function notSynchronisedBetweenEsAndMysql(int $batchSize): iterable
```

File: src/Akeneo/Pim/Enrichment/Bundle/Storage/ElasticsearchAndSql/ProductAndProductModel/IndexProductAndProductModel.php [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+                        return $updateDateInDb != $updateDateInIndex;
+                    }
+                )
+            );
+
+            yield $diff;
+        }
+    }
+
+    public function getAllProductUuids(int $batchSize): iterable
```
Comment (ahocquard) [inline src/Akeneo/Pim/Enrichment/Bundle/Storage/ElasticsearchAndSql/ProductAndProductModel/IndexProductAndProductModel.php:91]:
```suggestion
    public function all(int $batchSize): iterable
```

File: src/Akeneo/Pim/Enrichment/Bundle/Storage/ElasticsearchAndSql/ProductAndProductModel/IndexProductAndProductModel.php [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+                return;
+            }
+
+            $lastUuidAsBytes = end($rows);
+
+            yield array_map(fn (string $uuid): UuidInterface => Uuid::fromBytes($uuid), $rows);
+        }
+    }
+
+    public function getExistingProductUuids(array $identifiers): array
```
Comment (ahocquard) [inline src/Akeneo/Pim/Enrichment/Bundle/Storage/ElasticsearchAndSql/ProductAndProductModel/IndexProductAndProductModel.php:124]:
```suggestion
    public function existingAmong(array $identifiers): array
```

File: tests/back/Pim/Enrichment/Integration/Storage/ElasticsearchAndSql/ProductAndProductModel/GetAllProductIntegration.php [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+use Akeneo\Test\Integration\TestCase;
+use Ramsey\Uuid\Lazy\LazyUuidFromString;
+
+/**
+ * @copyright 2022 Akeneo SAS (https://www.akeneo.com)
+ * @license   https://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class GetAllProductIntegration extends TestCase
+{
+    public function setUp(): void
```
Comment (mmetayer) [inline tests/back/Pim/Enrichment/Integration/Storage/ElasticsearchAndSql/ProductAndProductModel/GetAllProductIntegration.php:17]:
```suggestion
    protected function setUp(): void
```

File: tests/back/Pim/Enrichment/Integration/Storage/ElasticsearchAndSql/ProductAndProductModel/GetAllProductIntegration.php [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+use Akeneo\Pim\Enrichment\Bundle\Storage\ElasticsearchAndSql\ProductAndProductModel\GetAllProduct;
+use Akeneo\Test\Integration\TestCase;
+use Ramsey\Uuid\Lazy\LazyUuidFromString;
+
+/**
+ * @copyright 2022 Akeneo SAS (https://www.akeneo.com)
+ * @license   https://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class GetAllProductIntegration extends TestCase
```
Comment (mmetayer) [inline tests/back/Pim/Enrichment/Integration/Storage/ElasticsearchAndSql/ProductAndProductModel/GetAllProductIntegration.php:15]: final

File: tests/back/Pim/Enrichment/Integration/Storage/ElasticsearchAndSql/ProductAndProductModel/GetAllProductIntegration.php [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+{
+    public function setUp(): void
+    {
+        parent::setUp();
+
+        $this->get('akeneo_integration_tests.fixture.loader.product_and_product_model_with_removed_attribute')->load();
+    }
+
+    public function test_it_get_all_product_uuids(): void
```
Comment (mmetayer) [inline tests/back/Pim/Enrichment/Integration/Storage/ElasticsearchAndSql/ProductAndProductModel/GetAllProductIntegration.php:24]:
```suggestion
    public function test_it_gets_all_product_uuids(): void
```

File: tests/back/Pim/Enrichment/Integration/Storage/ElasticsearchAndSql/ProductAndProductModel/GetAllProductIntegration.php [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    {
+        $productUuids = $this->getAllProduct()->getUuids(2);
+        $result = array(...$productUuids);
```
Comment (mmetayer) [inline tests/back/Pim/Enrichment/Integration/Storage/ElasticsearchAndSql/ProductAndProductModel/GetAllProductIntegration.php:27]:
```suggestion
        $result = [...$productUuids];
```

**Review by mmetayer — APPROVED** [review]
I let a lot of comments, but they're mostly naming suggestions and phpdoc/assertions
Good job! :rocket:

**Pull request merged**